### PR TITLE
Doc/add assemblers docs

### PR DESCRIPTION
**Pull request by J15525**
Updated docs for the assemblers module, including the weak forms.

@dccabanas, I haven't updated the 1-form and Maxwell eigenvalue solve functions that perform hierarchical refinements. Do you think they are useful as is? Or should they be turned into an example and removed?

### CODE REVIEW

**Comment by dccabanas** [comment]
> Updated docs for the assemblers module, including the weak forms.
> 
> @dccabanas, I haven't updated the 1-form and Maxwell eigenvalue solve functions that perform hierarchical refinements. Do you think they are useful as is? Or should they be turned into an example and removed?

I think for now we can ignore them, since the hierarchical module is being reworked and everything will be different after that.

File: src/Assemblers/WeakFormulations/WeakFormInputs.jl [hunk 1]
```diff
@@ -37,6 +25,39 @@ Container for test and trial spaces, and forcing terms to be used in a weak-form
...
+- `forcings::F`: The forcing terms for the weak-formulation, possibly nothing.
+
+# Type parameters
+- `manifold_dim::Int`: The dimension of the manifold where the weak-formulation is defined.
+- `TeF`: The type of the tupe of test forms. Each entry should be a subtype of
+    [`Forms.AbstractFormSpace`](@ref).
+- `TrF`: The type of the tupe of trial forms. Each entry should be a subtype of
+    [`Forms.AbstractFormSpace`](@ref).
+- `F`: The type of the tupe of forcing terms. Each entry should be a subtype of
+    [`Forms.AbstractFormField`](@ref).
```
Comment (dccabanas) [inline src/Assemblers/WeakFormulations/WeakFormInputs.jl:60]: `tupe` should be `tuple`

File: docs/src/Design/Modules/Assemblers.md [hunk 2]
```diff
@@ -1,6 +1,175 @@
...
+The assembly module contains two important pieces for assembly: the weak formulation and the assembly routine itself.
+The weak formulation contains all the terms that need to be assembled, the assembler creates the final system matrix.
+
+## [Weak Formulations](@id AssemblersWeakForms)
+A weak formulation describes the problem in detail, see [The Finite Element Method (FEM)](@ref) for more information.
+It consist of a few key ingredients: the given data (forcings, boundary conditions, etc.), the trail spaces, the test spaces, and the equations.
+In `Mantis`, there are two structures that are used to represent all this data: `WeakFormInputs` and `WeakForm`.
+
+### Representing problem data using `WeakFormInputs`
+The data, trail spaces, and test spaces are represented by the following object.
```
Comment (dccabanas) [inline docs/src/Design/Modules/Assemblers.md]: `trail` should be `trial`

File: docs/src/Design/Modules/Assemblers.md [hunk 3]
```diff
@@ -1,6 +1,175 @@
...
 
-## All docstrings from Mantis.Assemblers
-```@autodocs
-Modules = [Mantis.Assemblers]
+The assembly module contains two important pieces for assembly: the weak formulation and the assembly routine itself.
+The weak formulation contains all the terms that need to be assembled, the assembler creates the final system matrix.
+
+## [Weak Formulations](@id AssemblersWeakForms)
+A weak formulation describes the problem in detail, see [The Finite Element Method (FEM)](@ref) for more information.
+It consist of a few key ingredients: the given data (forcings, boundary conditions, etc.), the trail spaces, the test spaces, and the equations.
```
Comment (dccabanas) [inline docs/src/Design/Modules/Assemblers.md]: `consist` should be `consists`

File: docs/src/Design/Modules/Assemblers.md [hunk 4]
```diff
@@ -1,6 +1,175 @@
...
+
+### Example: setting up a weak form.
+As an example, consider the mixed ``n``-form Hodge-Laplacian in 3D also treated in [Finite Element Exterior Calculus (FEEC)](@ref).
+The weak formulation is written as:
+Given ``f \in L^2(\Omega)``, find ``(\sigma_h, u_h) \in V_h^{2} \times V_h^3`` such that
+```math
+    \begin{equation}
+        \begin{split}
+            (\sigma_h, \tau_h) - (u_h, d \tau_h) &= 0 \quad \forall \tau_h \in V_h^{2},\\
+            (d \sigma_h, v_h) &= (f, v_h) \quad \forall v_h \in V_h^3.
```
Comment (dccabanas) [inline docs/src/Design/Modules/Assemblers.md]: Missing commas before `\quad \forall`

File: docs/src/Design/Modules/Assemblers.md [hunk 5]
```diff
@@ -1,6 +1,175 @@
...
+        \end{split}
+    \end{equation}
+```
+We first create a `WeakFormInputs` object containing the test and trail spaces (which are the same in this case, so we only have to include them once), and the forcing (which we set to return one here for simplicity).
```
Comment (dccabanas) [inline docs/src/Design/Modules/Assemblers.md]: `trail` should be `trial`

File: docs/src/Design/Modules/Assemblers.md [hunk 6]
```diff
@@ -1,6 +1,175 @@
...
+```
+So, continuing the previous example, we can assemble the our problem by simply calling
+```@repl nFormHodgeLaplacianWeakForm
+A, b = Assemblers.assemble(weak_form);
+```
+This gives us the left-hand side matrix `A` and right-hand side vector `b`. 
+You can solve this with a solver of your choosing.
+Here we pick Julia's backslash operator.
+We can obtain the final solution as a `FormField` using the `build_form_fields` helper.
+This helper takes the trail spaces and full solution vector as inputs, and combines the spaces and coefficients into the final solution objects.
```
Comment (dccabanas) [inline docs/src/Design/Modules/Assemblers.md]: `trail` should be `trial`

File: docs/src/Design/Modules/Assemblers.md [hunk 7]
```diff
@@ -1,6 +1,175 @@
...
+solve_maxwell_eig
+```
+:::
+
+## Assembling a weak formulation
+In `Mantis`, the actual assembly routine can be used by simply calling the following function.
+```@docs
+assemble
+```
+So, continuing the previous example, we can assemble the our problem by simply calling
```
Comment (dccabanas) [inline docs/src/Design/Modules/Assemblers.md]: `the our` should be `our`

**Review by dccabanas — CHANGES_REQUESTED** [review]
Nice work! This is very clear and gives good information about the module.

PS: I noticed there is also an incorrect extra set of ` ``` ` in `Assemblers.add_bc!` when the arguments are two abstract matrices. This is from a previous change, but we should also correct that here.

**Comment by dccabanas** [comment]
@J15525 I noticed that in `assemble` we were indexing expression tuples, instead of looping over them. I made a small change to fix this, and there is a slight performance gain. Therefore, I'm including that as a commit here.

This is a benchmark for the example you added in the assembler doc page.

**Old**
```
BenchmarkTools.Trial: 10 samples with 1 evaluation per sample.
 Range (min … max):  430.711 ms … 575.400 ms  ┊ GC (min … max):  2.10% … 24.04%
 Time  (median):     542.066 ms               ┊ GC (median):    24.98%
 Time  (mean ± σ):   535.727 ms ±  38.963 ms  ┊ GC (mean ± σ):  22.94% ±  7.18%

  ▁                                         ▁▁  █▁ ▁ ▁ ▁      ▁
  █▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁██▁▁██▁█▁█▁█▁▁▁▁▁▁█ ▁
  431 ms           Histogram: frequency by time          575 ms <

 Memory estimate: 841.81 MiB, allocs estimate: 133009.
```

**New**
```
BenchmarkTools.Trial: 10 samples with 1 evaluation per sample.
 Range (min … max):  406.191 ms … 571.128 ms  ┊ GC (min … max):  5.18% … 24.30%
 Time  (median):     549.676 ms               ┊ GC (median):    24.66%
 Time  (mean ± σ):   527.201 ms ±  55.845 ms  ┊ GC (mean ± σ):  21.30% ±  8.93%

  ▁            ▁                                  ▁▁  ▁█ ▁  ▁ ▁
  █▁▁▁▁▁▁▁▁▁▁▁▁█▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁██▁▁██▁█▁▁█▁█ ▁
  406 ms           Histogram: frequency by time          571 ms <

 Memory estimate: 835.58 MiB, allocs estimate: 131907.
```

**Comment by J15525** [comment]
> Nice work! This is very clear and gives good information about the module.
> 
> PS: I noticed there is also an incorrect extra set of `` ``` `` in `Assemblers.add_bc!` when the arguments are two abstract matrices. This is from a previous change, but we should also correct that here.

Thanks! I fixed the typos (I see I managed to go for trail again 😬), including the extra `` ``` ``.

**Comment by J15525** [comment]
> @J15525 I noticed that in `assemble` we were indexing expression tuples, instead of looping over them. I made a small change to fix this, and there is a slight performance gain. Therefore, I'm including that as a commit here.
> 
> This is a benchmark for the example you added in the assembler doc page.
> 
> **Old**
> 
> ```
> BenchmarkTools.Trial: 10 samples with 1 evaluation per sample.
>  Range (min … max):  430.711 ms … 575.400 ms  ┊ GC (min … max):  2.10% … 24.04%
>  Time  (median):     542.066 ms               ┊ GC (median):    24.98%
>  Time  (mean ± σ):   535.727 ms ±  38.963 ms  ┊ GC (mean ± σ):  22.94% ±  7.18%
> 
>   ▁                                         ▁▁  █▁ ▁ ▁ ▁      ▁
>   █▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁██▁▁██▁█▁█▁█▁▁▁▁▁▁█ ▁
>   431 ms           Histogram: frequency by time          575 ms <
> 
>  Memory estimate: 841.81 MiB, allocs estimate: 133009.
> ```
> 
> **New**
> 
> ```
> BenchmarkTools.Trial: 10 samples with 1 evaluation per sample.
>  Range (min … max):  406.191 ms … 571.128 ms  ┊ GC (min … max):  5.18% … 24.30%
>  Time  (median):     549.676 ms               ┊ GC (median):    24.66%
>  Time  (mean ± σ):   527.201 ms ±  55.845 ms  ┊ GC (mean ± σ):  21.30% ±  8.93%
> 
>   ▁            ▁                                  ▁▁  ▁█ ▁  ▁ ▁
>   █▁▁▁▁▁▁▁▁▁▁▁▁█▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁██▁▁██▁█▁▁█▁█ ▁
>   406 ms           Histogram: frequency by time          571 ms <
> 
>  Memory estimate: 835.58 MiB, allocs estimate: 131907.
> ```

Yeah, I suspected this module could do with some upgrades. I also noticed this module, amongst other, does not yet have inference tests. But this is a useful first step.

**Review by dccabanas — APPROVED** [review]
